KAFKA-4954: Request handler utilization quotas

rajinisivaram opened the pull request:
See KIP-124 (https://cwiki.apache.org/confluence/display/KAFKA/KIP-124+-+Request+rate+quotas) for details

rajinisivaram committed:
KAFKA-4954: Request quotas based on I/O and network thread utilization

junrao commented on core/src/main/scala/kafka/network/RequestChannel.scala:
```diff
@@ -164,6 +172,14 @@ object RequestChannel extends Logging {
+      // the user/client. Throttling is only performed when request handler thread usage
+      // is recorded, just before responses are queued for delivery.
+      recordNetworkThreadUsageCallback match {
+        case Some(record) => record(math.max(endTimeNanos - responseDequeueTimeNanos, 0))
```
Hmm, not sure if this is accurate to capture the network thread utilization. What we are recording is essentially the responseSendTime, which includes the time for waiting for the socket to be writable. That portion of the time actually doesn't tie up the network threads and shouldn't be accounted for in the request time. Also, this seems to only cover the network thread time for sending responses, not for reading requests (which could be significant for produce requests).

I was thinking that we may need to do the following. In Selector.pollSelectionKeys(), we will measure the time spent for reading/writing each KafkaChannel and propagate the time back to the caller. Then, we can account for both request/response time in network threads in SocketServer.

rajinisivaram replied:
@junrao Thank you, I have reworked the code for recording network thread time as you suggested.  At the moment, SSL/SASL handshake time is included in the time for the first request. Is that reasonable? I can clear the time after authentication if it is confusing.

junrao replied:
That seems fine for now. Perhaps we could add a comment in case we need to revisit in the future?

rajinisivaram replied:
Added comment in `SocketServer`.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -58,6 +59,8 @@ object ClientQuotaManagerConfig {
   val DefaultQuotaWindowSizeSeconds = 1
   // Purge sensors after 1 hour of inactivity
   val InactiveSensorExpirationTimeSeconds  = 3600
+  val QuotaThreadPercentDefault = Int.MaxValue.toDouble
```
QuotaThreadPercentDefault => QuotaRequestPercentDefault ?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -185,8 +204,9 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
+        val clientQuotaEntity = clientSensors.quotaEntity
         val clientMetric = metrics.metrics().get(clientRateMetricName(clientQuotaEntity.sanitizedUser, clientQuotaEntity.clientId))
-        throttleTimeMs = throttleTime(clientMetric, getQuotaMetricConfig(clientQuotaEntity.quota))
+        throttleTimeMs = throttleTime(clientMetric, getQuotaMetricConfig(clientQuotaEntity.quota), alreadyThrottledMs).round.toInt
```
It seems that toInt is redundant.

rajinisivaram replied:
Removed.

rajinisivaram replied:
Sorry, had to revert that since `toInt` was required for newer versions of scala.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -325,13 +354,13 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
     // Use the precise window used by the rate calculation
     val throttleTimeMs = difference / quota.bound * rateMetric.windowSize(config, time.milliseconds())
-    throttleTimeMs.round.toInt
+    math.max(0, throttleTimeMs.round.toLong - alreadyThrottledMs)
```
It seems that toLong is redundant.

rajinisivaram replied:
Removed.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -346,39 +375,54 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
   }
 
-  private def getThrottleTimeSensorName(quotaId: QuotaId): String = apiKey + "ThrottleTime-" + quotaId.sanitizedUser.getOrElse("") + ':' + quotaId.clientId.getOrElse("")
+  protected def measurableStat: MeasurableStat = new Rate()
```
Do we need to make this protected? It doesn't seems to be customized in the subclass and it doesn't seem that we can change it to anything other than Rate().

rajinisivaram replied:
Reverted.

junrao commented on core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala:
```diff
@@ -0,0 +1,56 @@
+import org.apache.kafka.common.metrics._
+import org.apache.kafka.common.utils.Time
+
+import scala.collection.JavaConverters._
```
unused import

rajinisivaram replied:
Removed.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -508,20 +519,23 @@ class KafkaApis(val requestChannel: RequestChannel,
-        val responseSend = response.toSend(responseStruct, throttleTimeMs, request.connectionId, request.header)
-        requestChannel.sendResponse(new RequestChannel.Response(request, responseSend))
+      def fetchResponseCallback(bandwidthThrottleTimeMs: Int) {
+        def sendResponseCallback(utilizationThrottleTimeMs: Int) {
```
utilizationThrottleTimeMs => requestThrottleTimeMs?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1130,15 +1158,18 @@ class KafkaApis(val requestChannel: RequestChannel,
-      trace("Sending leave group response %s for correlation id %d to client %s."
+      def createResponse(throttleTimeMs: Int): AbstractResponse = {
+        val response = new LeaveGroupResponse(throttleTimeMs, error)
+        trace(  "Sending leave group response %s for correlation id %d to client %s."
```
remove space after trace( ?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1312,4 +1349,48 @@ class KafkaApis(val requestChannel: RequestChannel,
+        quotaSensors,
+        nanosToPercentage(request.requestHandlerThreadTimeNanos),
+        sendResponseCallback,
+        bandwidthThrottleTimsMs)
```
Hmm, I am wondering if we really need to subtract bandwidthThrottleTimsMs from the throttle time. While the response is delayed for bandwidthThrottleTimsMs, some time has passed, which should bring down the value of the metric when we check the request quota. Then, naturally, this request will be delayed less for request quota violation.

rajinisivaram replied:
Yes, I wasn't sure which way to go. Either way, I think the throttle times will correct themselves over time. I have removed the subtraction.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -228,7 +236,7 @@ class KafkaApis(val requestChannel: RequestChannel,
     val controlledShutdownResponse = new ControlledShutdownResponse(controlledShutdownRequest.correlationId,
       Errors.NONE, partitionsRemaining)
-    requestChannel.sendResponse(new Response(request, new RequestOrResponseSend(request.connectionId, controlledShutdownResponse)))
+    sendResponseExemptThrottle(request, new Response(request, new RequestOrResponseSend(request.connectionId, controlledShutdownResponse)))
```
We throttle LeaderAndIsrRequest if it's unauthorized. Should we do the same thing here?

rajinisivaram replied:
In this case `authorizeClusterAction` method throws an exception and the response is sent from the exception handler in `handle`. All responses sent from the exception handler are throttled.

junrao replied:
Hmm, in this case, we probably only want to throttle if the exception is related to authorization. For any other exceptions, we should send an error immediately?

rajinisivaram replied:
Yes, `authorizeClusterAction` takes care of throttling for unauthorized request and all other paths including error path goes through this `sendResponseExemptThrottle` which does not perform throttling.

junrao reviewed (commented):
@rajinisivaram : Thanks for the patch. Made a pass and left some comments below.

rajinisivaram commented:
@junrao Thank you for the review, I have addressed the comments and left one question.

System tests passed here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/814/

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -65,6 +65,9 @@
     /* The v2 metadata request is the same as v1. An additional field for cluster id has been added to the v2 metadata response */
     public static final Schema METADATA_REQUEST_V2 = METADATA_REQUEST_V1;
 
+    /* The v3 metadata request is the same as v1 and v2. An additional field for throttle time has been added to the v3 metadata response */
```
Could we add a comment somewhere so that people know that if they want to add a non-internal request in the future, they would need to include the throttle time field?

rajinisivaram replied:
I have added unit test in `ApiKeysTest` that checks that all responses except those explicitly excluded contain a field named `throttle_time_ms`. There is a comment in the test as well to ensure that new requests either contain the field or are manually added to the test's exclusion list.

ijuma replied:
I wonder if there's a way to add this to `ResponseHeader` in a compatible way. It seems a bit annoying to have to add that to every response.

rajinisivaram replied:
@ijuma There are a few requests which don't have throttle time (ControlledShutdown, StopReplica etc. used for inter-broker and not by producer/consumer). So it made sense to add throttle time to individual requests. Error codes seem to be handled this way. But my main concern was compatibility. Even if we looked at request version to determine the response header version, there is still the issue of `ApiVersionsResponse` which needs to be handled. According to KIP-35, `ApiVersionResponse.ErrorCode is guaranteed to be the first int16 of the response for all future versions of ApiVersionRequest`. I didn't want to break that assumption.

ijuma replied:
Sorry for the delay on this one. So, one way to do this would be the following:

1. Add two fields to ResponseHeader _if_ the request version is higher than the version before this PR: error_code, throttle_time_ms (in this order)
2. Remove any top-level error_code in the new version of all affected responses
3. Remove throttle_time_ms from FetchResponse and ProduceResponse in the new version
4. throttle_time_ms is always 0 for requests that are never throttled

As part of this, we would also solve the issue that we currently have no way to return generic errors via the protocol. Since we are bumping the protocol version for so many requests, it seems like it would be a good opportunity to fix both issues at the same time. Is there a reason why this is a bad idea or would not work?

ijuma replied:
Discussed this with @junrao. The main challenge with this option is having the top level error field for every response. This would probably affect a lot of code:

1. We would need to handle this top level error code everywhere.
2. A bunch of protocols that currently have a top level error code would no longer have them, so a bunch of code would have to be updated as well.

So, it doesn't seem appropriate to do this as part of this KIP.

rajinisivaram replied:
@ijuma Thank you for looking into this.

junrao commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -505,14 +506,13 @@ private[kafka] class Processor(val id: Int,
-          buffer = receive.payload, startTimeMs = time.milliseconds, listenerName = listenerName,
-          securityProtocol = securityProtocol)
+          buffer = receive.payload, startTimeNanos = time.nanoseconds, networkThreadTimeNanos = networkThreadTimeNanos,
+          listenerName = listenerName, securityProtocol = securityProtocol)
```
This takes care of the time spent on receiving requests. We will need to do the same thing to track the time on sending responses.

rajinisivaram replied:
The time is accumulated in `Selector` and includes the full time spent for each channel in `pollSelectionKeys`. For each request, the accumulated time is used and the value is reset. This time includes the time for write of the previous response and the time for read of the current request. I have added a comment in the code.

rajinisivaram replied:
After writing the comment above, I realized that it sounds rather odd. So I have updated the code to record network thread time when request metrics are updated, so that receive+send are recorded together. This does also ensure that the last network time on each connection is recorded and it would work even if requests on a connection use different clientIds. Throttling is still performed only on subsequent requests.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -146,16 +157,15 @@ class KafkaApis(val requestChannel: RequestChannel,
+      } else {
+        val result = leaderAndIsrRequest.partitionStates.asScala.keys.map((_, Errors.CLUSTER_AUTHORIZATION_FAILED)).toMap
+        def createResponse(throttleTimeMs: Int): AbstractResponse = new LeaderAndIsrResponse(Errors.CLUSTER_AUTHORIZATION_FAILED, result.asJava)
+        sendResponseMaybeThrottle(request, createResponse)
```
The approach in the patch works. But one issue is that the implementor of all future requests will have to deal with throttling the responses directly.

Another approach is to do the throttling early in handle(). If a request needs to be delayed, we first throttle it and then hand it over to the specific request handler. This way, we just need to implement the request throttling logic in one place and all future requests don't have to be aware of it. We probably need to mark whether a request is at the cluster level so that we can throttle unauthorized internal requests too. Also, not sure if there is an easier way for the request handler thread to pick up the requests after throttling is done. If there is, this may be a simpler approach?

rajinisivaram replied:
@junrao Thank you for the review. 

It will be nice to handle throttling in a single place. However, handling of all new requests need to be aware of throttling, so that they add the throttling time to the response. The bigger issue is the exclusions. We need to authorize in a central place for `ClusterAction`. And worse, we need to handle `Produce` differently since we don't want to throttle until after the request is processed and the memory can be released. We probably want to throttle later for `Fetch` as well since we are relying on the bandwidth throttle time to reduce the request throttle time. We will need to record two more timestamps to take into account time spent before throttle in some cases and after throttle in others. And as you mentioned, we still want the request to be processed after throttling on the request handler thread. Taking all that into account, I am not  sure it is worthwhile to restructure the code to centralize the throttling logic.

I have updated `RequestQuotaTest` to ensure that unauthorized requests of all types are throttled. And also to check that all requests not explicitly excluded are throttled and return throttle time in response. This should catch any missing throttling in new requests.

Let me know if this is sufficient or whether I should try out the centralized approach.

ijuma replied:
If we could somehow add the throttle time in the response header, it would be easier to handle it in a generic way for the typical cases.

junrao reviewed (commented):
@rajinisivaram : Thanks for the updated patch. A few more comments.

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -24,13 +24,15 @@
     private final String id;
     private final TransportLayer transportLayer;
     private final Authenticator authenticator;
+    private final AtomicLong networkThreadTimeNanos;
```
Why is this an `AtomicLong` instead of a plain long?

rajinisivaram replied:
Because the value is updated on the network thread and read-and-reset on the request handler thread. I have added a comment.

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -326,6 +340,7 @@ private void pollSelectionKeys(Iterable<SelectionKey> selectionKeys,
             SelectionKey key = iterator.next();
             iterator.remove();
             KafkaChannel channel = channel(key);
+            long channelStartTimeNanos = recordTimePerConnection ? time.nanoseconds() : 0;
```
Hmm, it seems a bit unfortunate that we need to do a `System.nanoTime` per selection key. Have we done any measurements on the overhead?

rajinisivaram replied:
I couldn't see any way around it. I will kick off system test runs to see the impact.

junrao replied:
Our system test currently doesn't do perf validation well. It would be useful to just run ProducerPerformance and ConsumerPerformance and see if there is any noticeable degradation.

rajinisivaram replied:
@junrao @ijuma I ran `ProducerPerformance` and `ConsumerPerformance` on my laptop and didn't see any noticeable difference. This is the throughput in MB/s (average of three runs):

Test     (message size)    | trunk             |  with PR
-------------------------|--------------|------------
Producer (100 bytes)      | 158.68          | 160.18 
Producer (1000 bytes)    | 355.43         | 350.36
Consumer (100 bytes)    | 376.20          | 378.52 
Consumer (1000 bytes)  | 559.45          | 559.45

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -112,7 +113,16 @@ public Struct parseRequest(short version, ByteBuffer buffer) {
+        try {
+            return responseSchema(version).read(buffer);
+        } catch (SchemaException e) {
+            if (this == ApiKeys.API_VERSIONS && version != 0) {
```
I think we'd want to override `parseResponse` for `API_VERSIONS` only.

rajinisivaram replied:
Done.

ijuma commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -58,6 +59,8 @@ object ClientQuotaManagerConfig {
   // Purge sensors after 1 hour of inactivity
   val InactiveSensorExpirationTimeSeconds  = 3600
+  val QuotaRequestPercentDefault = Int.MaxValue.toDouble
+  val NanosToPercentagePerSecond = 100.0 / TimeUnit.SECONDS.toNanos(1)
```
Since we use `Ms`, should we not be using `Ns` instead of `Nanos`?

rajinisivaram replied:
Both `Ns` and `Nanos` are already used in the code. I chose `Nanos` everywhere in the PR since it stands out better from `Ms`, especially since I was changing some measurements from millis to nanos. But I am ok with switching to `Ns` if that is preferable.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -106,22 +106,33 @@ class KafkaApis(val requestChannel: RequestChannel,
+              requestChannel.closeConnection(request.processor, request)
+            response
+          }
+          sendResponseMaybeThrottle(request, createResponse)
```
Nit: the logic inside this catch has become a bit complicated. Can we perhaps extract methods? Also, something I was thinking about recently is that methods like `getErrorResponse` could throw an exception due to bugs. It would be nice for us not to leak connections in such cases.

rajinisivaram replied:
Have moved to a method. I think errors are propagated and logged. We won't close acks=0 connections if `getErrorResponse` threw an exception, but perhaps that is ok since that would still not be a leaked connection?

ijuma reviewed (commented):
Thanks for the PR. I left a few questions and comments.

rajinisivaram commented:
@ijuma Thank you for the review. I haven't moved throttle-time to the response header as explained in the previous comment. One alternative to making the code in `KafkaApis` neater would be to make `throttleTimeMs` a non-final field. That way the response could be created as it is now, and throttle time can just be updated in the response. I didn't want to do that initially because it makes the code inconsistent, but perhaps that is ok for this case?

junrao commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -564,10 +563,14 @@ private void handleApiVersionsResponse(List<ClientResponse> responses,
+                this.selector.close(node);
+                processDisconnection(responses, node, now);
+            } else {
+                nodesNeedingApiVersionsFetch.put(node, new ApiVersionsRequest.Builder((short) 0));
```
Perhaps we should only fall back to version 0 of the request if the error is UNSUPPORTED_VERSION? For other kinds of error, just disconnect?

rajinisivaram replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -164,6 +169,24 @@ public Send write() throws IOException {
+     * and read/reset by the broker's request handler thread.
+     */
+    public AtomicLong networkThreadTimeNanos() {
+        return networkThreadTimeNanos;
```
Instead of returning AtomicLong, could we just reset to 0 and return a long?

rajinisivaram replied:
AtomicLong is returned so that the the value can be updated from the I/O thread when a request is complete without propagating `KafkaChannel` to the request handling code.

junrao replied:
Got it. An alternative is to call networkThreadTimeNanos() in SocketServer.processNewResponses() and processCompletedSends(). Then we can just reset and return the value, which is easier to understand?

rajinisivaram replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -43,7 +44,12 @@
+    API_VERSIONS(18, "ApiVersions") {
+        @Override
+        public Struct parseResponse(short version, ByteBuffer buffer) {
+            return parseResponse(version, buffer, (short) 0);
```
Hmm, not sure why we need this. It seems that the client should always use the requested version to parse the response of API_VERSIONS?

rajinisivaram replied:
If client sends ApiVersionsRequest with a higher version that client supports, broker responds with a version 0 response that indicates unsupported version.

junrao replied:
Thanks for the explanation. Could we add that as comment in the code?

junrao commented on core/src/main/scala/kafka/network/RequestChannel.scala:
```diff
@@ -57,14 +58,16 @@ object RequestChannel extends Logging {
+    @volatile var responseCompleteTimeNanos = -1L
+    @volatile var responseDequeueTimeNanos = -1L
+    @volatile var apiRemoteCompleteTimeNanos = -1L
+    @volatile var recordNetworkThreadTimeCallback: Option[Long => Unit] = None
```
This makes things a bit more complicated. I was thinking of the following. In updateRequestMetrics(), we remember networkThreadTime as previousNetworkThreadTime. In KafkaApis, we can just add previousNetworkThreadTime to the throttler. That way we don't need this callback. Will that be better?

rajinisivaram replied:
Hmm... Network thread time needs to be accumulated against the (user, client-id) and needs to include the time for the sending the response. The callback avoids having to propagate (user, client-id).

junrao replied:
Got it. This is fine then.

junrao commented on core/src/main/scala/kafka/network/RequestChannel.scala:
```diff
@@ -122,26 +125,33 @@ object RequestChannel extends Logging {
     trace("Processor %d received request : %s".format(processor, requestDesc(true)))
 
+    def requestThreadTimeNanos = {
+      val localCompleteTimeNanos = if (apiLocalCompleteTimeNanos == -1L) Time.SYSTEM.nanoseconds else apiLocalCompleteTimeNanos
```
It seems that we should update the instance level localCompleteTimeNanos instead of a local one?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -129,9 +132,9 @@ case class QuotaEntity(quotaId: QuotaId, sanitizedUser: String, clientId: String
+class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
                          private val metrics: Metrics,
-                         private val apiKey: QuotaType,
+                         private val quotaType: QuotaType,
```
apiKey in the comment needs to be changed accordingly.

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -197,6 +207,15 @@ final class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
 
   /**
+   * Records that a user/clientId changed some metric being throttled without checking for
+   * quota violation. The aggregate value will subsequently used for throttling when the
```
will subsequently used => will subsequently be used

rajinisivaram replied:
Done.

junrao reviewed (commented):
@rajinisivaram : Thanks for the patch. Left a few more comments. Could you rebase? Also, https://shipilev.net/blog/2014/nanotrusting-nanotime/ seems to indicate that nanotime is a scalability bottleneck. Could you run the perf test with multiple producer/consumer instances?

ijuma commented:
When it comes to performance tests, it would also be good to run them on AWS as VMs like Xen often have an impact of time-related methods. For example: https://blog.packagecloud.io/eng/2017/03/08/system-calls-are-much-slower-on-ec2/

rajinisivaram commented:
@junrao @ijuma Thank you for the reviews. I have rebased.

I have performance tests on my laptop with 8 threads and see no noticeable difference.
- Producer (100 byte messages): 161.92 MB/s (before) and 164.4 MB/s(after)
- Consumer (100 byte messages) : 168.27 MB/se (before) and 173.5 MB/s (after)

I don't have an account to run the tests on AWS, but I will start a system test run to compare.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -1082,15 +1154,22 @@
+    public static final Schema API_VERSIONS_RESPONSE_V1 = new Schema(
+            new Field("error_code", INT16, "Error code."),
+            new Field("api_versions", new ArrayOf(API_VERSIONS_V0), "API versions supported by the broker."),
+            newThrottleTimeField());
```
Could we consistently add newThrottleTimeField() as the first field?

rajinisivaram replied:
According to KIP-35,
> ApiVersionResponse.ErrorCode is guaranteed to be the first int16 of the response for all future versions of ApiVersionRequest and is to be used to indicate that the client's ApiVersionRequest with version X (greater than 0) is not supported by the broker and the client should revert to version 0.

The Java clients don't rely on this, but just in case some other clients do, I have left `error_code` as the first field for `ApiVersionsResponse`.

junrao replied:
Ok, could we add a comment that error_code has to be the first field in ApiResponse?

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -1265,6 +1361,7 @@
                     "The partition")
             );
     public static final Schema OFFSET_FOR_LEADER_EPOCH_RESPONSE_V0 = new Schema(
+            newThrottleTimeField(),
```
OFFSET_FOR_LEADER_EPOCH_RESPONSE is an inter broker request. So, we shouldn't add a throttle field.

rajinisivaram replied:
Fixed.

junrao commented on core/src/main/scala/kafka/network/RequestChannel.scala:
```diff
@@ -122,26 +125,33 @@ object RequestChannel extends Logging {
+      val apiThrottleTime = nanosToMs(responseCompleteTimeNanos - apiRemoteCompleteTimeNanos)
+      val responseQueueTime = nanosToMs(responseDequeueTimeNanos - responseCompleteTimeNanos)
+      val responseSendTime = nanosToMs(endTimeNanos - responseDequeueTimeNanos)
+      val totalTime = nanosToMs(endTimeNanos - startTimeNanos)
```
It might be useful to report all the time still as ms, but up to micro sec level accuracy now that we track with nanosec.

rajinisivaram replied:
The yammer metrics `Histogram` class that tracks time only takes long and not double. Hence the millisecond value is used.

junrao replied:
I was actually referring to line 169 where we log the time components in trace logging. It's useful to see more precise time there since sometimes the time may take less than 1ms.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -346,39 +368,54 @@ final class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
   }
 
-  private def getThrottleTimeSensorName(quotaId: QuotaId): String = apiKey + "ThrottleTime-" + quotaId.sanitizedUser.getOrElse("") + ':' + quotaId.clientId.getOrElse("")
+  private def measurableStat: MeasurableStat = new Rate()
```
For request quota, since we are collecting request time in nanosecs already, it will be useful to create a Rate with nanosec as the time unit. This will make the measurement more accurate.

rajinisivaram replied:
For request quota, values are recorded as double, so even though they use millisecond as unit to be consistent, they have higher precision. Isn't that sufficient?

junrao replied:
Thanks. That should be enough then.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -516,20 +501,23 @@ class KafkaApis(val requestChannel: RequestChannel,
+          val responseSend = response.toSend(responseStruct, bandwidthThrottleTimeMs + requestThrottleTimeMs, request.connectionId, request.header)
+          requestChannel.sendResponse(new RequestChannel.Response(request, responseSend))
+        }
+        sendResponseMaybeThrottle(request, request.header.clientId, sendResponseCallback)
```
Hmm, this can be a bit tricky. Fetch requests from the follower are considered internal and shouldn't be throttled.

rajinisivaram replied:
Oops, you are right. Fixed.

junrao commented on core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:
```diff
@@ -136,6 +141,28 @@ abstract class BaseQuotaTest extends IntegrationTestHarness {
+    val endTimeMs = System.currentTimeMillis + 10000
+    var throttled = false
+    while (!throttled && System.currentTimeMillis < endTimeMs) {
+      consumer.poll(100)
```
Hmm, when there is no data, the consumer will wait for the timeout. So, not sure if this is enough to trigger the throttling. We probably need to either set a low maxWait in consumer config and set the quota to be really low.

rajinisivaram replied:
Consumers in this test are configured with `fetch.max.wait.ms=0`. The quota is set very small as well to trigger throttling.

junrao commented on core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:
```diff
@@ -0,0 +1,428 @@
+      var correlationId: Int = 1
+      for (apiKey <- RequestQuotaTest.ClientActions) {
+        correlationId += 1
+        val response = requestResponse(socket, unthrottledClientId, correlationId, requestBuilder(apiKey))
```
Hmm, not sure where the test is.

rajinisivaram replied:
Oops, fixed.

junrao commented on core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:
```diff
@@ -0,0 +1,428 @@
+        case ApiKeys.FETCH =>
+          val partitionMap = new LinkedHashMap[TopicPartition, requests.FetchRequest.PartitionData]
+          partitionMap.put(tp, new requests.FetchRequest.PartitionData(0, 0, 100))
+          requests.FetchRequest.Builder.forReplica(ApiKeys.FETCH.latestVersion, 5000, 100, Int.MaxValue, partitionMap)
```
Why is the replicaId 5000? That indicates it's from a follower. Also, I am wondering if 100 maxWait is enough to trigger throttling.

rajinisivaram replied:
Copy-paste error. Fixed and set maxWait to zero.

junrao reviewed (commented):
@rajinisivaram : Thanks for the new patch. Made another pass and left some more comments.

ijuma commented:
By the way, since the additional `System.nanoTime` call per selection key only happens if the new config is enabled, I think it's OK to merge this and do additional performance testing later. I think this is one of the cases where issues only show themselves when the number of threads and CPUs is larger than one can test on a laptop.

ijuma commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -183,8 +183,8 @@ public void addNetworkThreadTimeNanos(long nanos) {
      * This is an AtomicLong because the value is updated by the network thread
      * and read/reset by the broker's request handler thread.
      */
```
I think the AtomicLong comment should be moved to the field.

rajinisivaram replied:
Done.

rajinisivaram commented:
@ijuma Yes, agree that it is difficult to assess the impact of the change on a laptop.
@junrao I have addressed most of the issues. Left comments for the others.

Many thanks for the reviews.

ijuma commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -529,11 +530,29 @@ private[kafka] class Processor(val id: Int,
+      request.recordNetworkThreadTimeCallback match {
+        case Some(record) => record(openOrClosingChannel.getAndResetNetworkThreadTimeNanos())
+        case None =>
+      }
```
Is there a reason why we don't replace the 4 lines above with:

```scala
    request.recordNetworkThreadTimeCallback.foreach(record => record(openOrClosingChannel.getAndResetNetworkThreadTimeNanos())
```

rajinisivaram replied:
@ijuma Thank you, done.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1367,11 +1392,81 @@ class KafkaApis(val requestChannel: RequestChannel,
+        response
+      }
+      error("Error when handling request %s".format(request.body[AbstractRequest]), e)
+      sendResponseMaybeThrottle(request, createResponse)
```
Here, we throttle independent of the exception type. Perhaps, it's better to only engage in throttling if the exception is ClusterAuthorizationException?

rajinisivaram replied:
Fixed to throttle only for `ClusterAuthorizationException` for broker-only requests.

junrao reviewed (commented):
@rajinisivaram : Thanks for the updated patch. A few more followup comments. Also, (1) Could you run perf test with a really high request throttling value and see if there is no degradation? (2) Could you run our system tests?

rajinisivaram commented:
@junrao Thank you for the review.
I have run the performance tests on my laptop with high request quota configured. Again, there was no noticeable difference.
Producer (100 bytes): trunk: 151.33 MB/s   requestQuota: 155.78 MB/s
Consumer (100 bytes): trunk: 332.32 MB/s   requestQuota: 339.64 MB/s

Also started a system test run with the latest level of code: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/855/

rajinisivaram commented:
System tests have passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/855/

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -25,35 +26,43 @@
+    ADD_PARTITIONS_TO_TXN(24, "AddPartitionsToTxn", false),
+    ADD_OFFSETS_TO_TXN(25, "AddOffsetsToTxn", false),
+    END_TXN(26, "EndTxn", false),
+    WRITE_TXN_MARKERS(27, "WriteTxnMarkers", false),
```
This is an inter-broker request as well and clusterAction should be true.

rajinisivaram replied:
Fixed.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -1387,6 +1486,7 @@
     );
 
     public static final Schema WRITE_TXN_MARKERS_RESPONSE_V0 = new Schema(
+            newThrottleTimeField(),
```
This is an inter-broker request. So, no throttling needed.

rajinisivaram replied:
Fixed.

junrao commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -529,11 +530,25 @@ private[kafka] class Processor(val id: Int,
+    val channel = selector.channel(channelId)
+    val openOrClosingChannel = if (channel != null) channel else selector.closingChannel(channelId)
+    if (openOrClosingChannel != null)
+      request.recordNetworkThreadTimeCallback.foreach(record => record(openOrClosingChannel.getAndResetNetworkThreadTimeNanos()))
```
Perhaps it's better to pass the networkThread time to request.updateMetrics() and call the recordNetworkThreadTimeCallback there?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -453,6 +454,7 @@ private[kafka] class Processor(val id: Int,
     var curr = requestChannel.receiveResponse(id)
     while (curr != null) {
       try {
+        recordNetworkThreadTimeNanos(curr.request.connectionId, curr.request)
```
Hmm, we probably don't want to call recordNetworkThreadTimeNanos here since it will be called in processCompleteSends() again. Instead, it seems that we want to call recordNetworkThreadTimeNanos() in all places where we call request.updateRequestMetrics().

rajinisivaram replied:
Fixed.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:
```diff
@@ -24,13 +24,17 @@
     private final TransportLayer transportLayer;
     private final Authenticator authenticator;
+    // Tracks time spent in the network thread. This is an AtomicLong since it is updated
+    // by the network thread and read and reset by the broker's request handler thread.
```
It's actually read and reset by the broker's network thread.

rajinisivaram replied:
Yes, you are right. Replaced with `long` and updated comment.

junrao commented on core/src/main/scala/kafka/network/RequestChannel.scala:
```diff
@@ -122,26 +123,33 @@ object RequestChannel extends Logging {
+      val apiRemoteTime = nanosToMs(apiRemoteCompleteTimeNanos - apiLocalCompleteTimeNanos)
+      val apiThrottleTime = nanosToMs(responseCompleteTimeNanos - apiRemoteCompleteTimeNanos)
+      val responseQueueTime = nanosToMs(responseDequeueTimeNanos - responseCompleteTimeNanos)
+      val responseSendTime = nanosToMs(endTimeNanos - responseDequeueTimeNanos)
```
For the trace logging in line 177, could we report all the time still as ms, but up to micro sec level accuracy?

rajinisivaram replied:
Done.

junrao commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -50,7 +50,10 @@ class KafkaRequestHandler(id: Int,
-          val idleTime = time.nanoseconds - startSelectTime
+          val endTime = time.nanoseconds
+          if (req != null)
+            req.requestDequeueTimeNanos = endTime
```
Given this, should we just remove line 65?

rajinisivaram replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:
```diff
@@ -0,0 +1,425 @@
+    }
+  }
+
+  def submitTest(apiKey: ApiKeys, test: () => Unit) {
```
Could we make this and a few other methods in the class private?

rajinisivaram replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:
```diff
@@ -0,0 +1,425 @@
+    } finally {
+      socket.close()
+    }
+    waitAndCheckResults()
```
Hmm, is the test added? I don't see the code for submitTest that checks the throttling field in the response.

rajinisivaram replied:
Sorry, had forgotten the tests, fixed now.

junrao reviewed (commented):
@rajinisivaram : Thanks for the updated patch. Left a few more more comments. Also, could you rebase?

rajinisivaram commented:
@junrao Thank you for the reviews. Have rebased and addressed the comments.

junrao commented:
@rajinisivaram : Thanks for the latest patch. LGTM.
